mage_integrations/mage_integrations/sources/oracledb/__init__.py

```diff
@@ -0,0 +1,316 @@
+    def host(self) -> str:
+        return self.config['host']
+
+    @property
+    def port(self) -> str:
+        return self.config['port']
+
+    @property
+    def service(self) -> str:
+        return self.config['service']
+
+    @property
+    def user(self) -> str:
+        return self.config['user']
+
+    @property
+    def password(self) -> str:
+        return self.config['password']
+
+    def nullable_column(self, col_name, col_type, pks_for_table):
+        if col_name in pks_for_table:
+            return [col_type]
+        else:
+            return ['null', col_type]
+
+    def get_database_name(self, connection):
+        cur = connection.cursor()
+        rows = cur.execute("SELECT name FROM v$database").fetchall()
+        return rows[0][0]
+
+    def produce_column_metadata(
+            self, connection, table_info, table_schema, table_name, pk_constraints, column_schemas, cols):
+        mdata = {}
+
+        table_pks = pk_constraints.get(table_schema, {}).get(table_name, [])
+
+        # NB> sadly, some system tables like XDB$STATS have P constraints
+        # for columns that do not exist so we must protect against this
+        table_pks = list(filter(
+            lambda pk: column_schemas.get(pk, Schema(None)).type is not None, table_pks))
+
+        database_name = self.get_database_name(connection)
+
+        metadata.write(mdata, (), 'table-key-properties', table_pks)
+        metadata.write(mdata, (), 'schema-name', table_schema)
+        metadata.write(mdata, (), 'database-name', database_name)
+
+        if table_schema in table_info and table_name in table_info[table_schema]:
+            metadata.write(mdata, (), 'is-view', table_info[table_schema][table_name]['is_view'])
+
+            row_count = table_info[table_schema][table_name].get('row_count')
+
+            if row_count is not None:
+                metadata.write(mdata, (), 'row-count', row_count)
+
+        for c in cols:
+            c_name = c.column_name
+            # Write the data_type or "None" when the column has no datatype
+            metadata.write(mdata, ('properties', c_name), 'sql-datatype', (c.data_type or "None"))
+            if column_schemas[c_name].type is None:
+                mdata = metadata.write(mdata, ('properties', c_name), 'inclusion', 'unsupported')
+                mdata = metadata.write(mdata, ('properties', c_name), 'selected-by-default', False)
+            elif c_name in pk_constraints.get(table_schema, {}).get(table_name, []):
+                mdata = metadata.write(mdata, ('properties', c_name), 'inclusion', 'automatic')
+                mdata = metadata.write(mdata, ('properties', c_name), 'selected-by-default', True)
+            else:
+                mdata = metadata.write(mdata, ('properties', c_name), 'inclusion', 'available')
+                mdata = metadata.write(mdata, ('properties', c_name), 'selected-by-default', True)
+
+        return mdata
+
+    def schema_for_column(self, c, pks_for_table):
```

Comment:
I think your implementation can be simplified a lot if you inherit the sql source.
You can just override the build_discover_query method like this https://github.com/mage-ai/mage-ai/blob/master/mage_integrations/mage_integrations/sources/mssql/__init__.py#L29-L54

Reply:
Thank you for the suggestion Xiaoyou! It is a much concise solution. Updated with overwriting of `build_discover_query` function.